smart_open/transport.py

```diff
@@ -63,29 +72,29 @@ def get_transport(scheme):
     This submodule must have been previously registered via :func:`register_transport`.
 
     """
+    global _ERRORS
     expected = SUPPORTED_SCHEMES
     readme_url = 'https://github.com/RaRe-Technologies/smart_open/blob/master/README.rst'
     message = (
         "Unable to handle scheme %(scheme)r, expected one of %(expected)r. "
         "Extra dependencies required by %(scheme)r may be missing. "
         "See <%(readme_url)s> for details." % locals()
     )
-    try:
-        submodule = _REGISTRY[scheme]
-    except KeyError as e:
-        raise NotImplementedError(message) from e
-    else:
-        return submodule
-
-
-register_transport(smart_open.local_file)
-register_transport('smart_open.azure')
-register_transport('smart_open.gcs')
-register_transport('smart_open.hdfs')
-register_transport('smart_open.http')
-register_transport('smart_open.s3')
-register_transport('smart_open.ssh')
-register_transport('smart_open.webhdfs')
+    if scheme in _REGISTRY:
+        return _REGISTRY[scheme]
+    if scheme in _ERRORS:
+        raise ImportError(_ERRORS[scheme])
+    raise NotImplementedError(message)
+
+
+register_transport(smart_open.local_file, ["file"])
```

Comment:
Why is it necessary to specify the second parameter here? It sounds like the sort of thing the submodules should know how to handle themselves.

Reply:
Yeah, that would normally be the case if you imported the submodule without error. But consider if you tried to open a blob (`gs://foo/bar`) without the GCS libraries installed. When the import error is raised from `gcs.py`, the `SCHEME/S` cannot be referenced in `transport.py`.

So without specifying the schemes here, when the open function is called for a GCS blob without the GCS libraries installed, the gs scheme isn't in the known list and transport has to show the generic "please look at this readme file" error.

Reply:
I removed this code by handling the ImportError in the various transports and setting a module attribute `MISSING_DEPS=True`. Transport is then able to inspect the schemes even if they have missing packages.